File: Leetcode/0075-SortColors/cpp/solution.cpp
class Solution {
 public:
  void sortColors(vector<int>& nums) {
    // i: next writing position for elements smaller than mid
    int i = 0;
    // j: scanning position
    int j = 0;
    // k: next writing position for elements greater than mid
    int k = nums.size() - 1;
    // Index for scanning the array
    while (j <= k) {
      if (nums[j] < 1) {
        std::swap(nums[i], nums[j]);
        i++;
        j++;
      } else if (nums[j] > 1) {
        std::swap(nums[j], nums[k]);
        k--;
      } else {
        j++;
      }
    }
  }
};

File: Leetcode/0075-SortColors/go/solution.go
func sortColors(nums []int)  {
    n := len(nums)
    i, j, k := 0, 0, n-1
    for j <= k {
        if nums[j] == 0 {
            nums[i], nums[j] = nums[j], nums[i]
            j++
            i++
        } else if nums[j] == 2 {
            nums[j], nums[k] = nums[k], nums[j]
            k--
        } else {
            j++
        }
    }
}